docs/proposals/11-scheduling.md

```diff
@@ -0,0 +1,162 @@
+---
+title: Reservation-Based Scheduling
+
+iep-number: 11
+
+creation-date: 2024-09-30
+
+status: draft
+
+authors:
+
+- "@lukasfrank"
+
+
+reviewers:
+
+- "@afritzler"
+- "@balpert89"
+
+---
+
+# IEP-11: Reservation-Based Scheduling
+
+## Table of Contents
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+- [Alternatives](#alternatives)
+
+## Summary
+Scheduling resources is a central process to increase utilization of cloud infrastructure. It not only involves the 
+process to find any feasible `MachinePool` but rather, if possible, to find a `MachinePool` where the `Machine` can 
+be materialized.
+
+
+## Motivation
+
+With `ironcore` a multi-level hierarchy can be built to design availability/failure zones. `Machine`s can be created 
+on every layer, which are subsequently scheduled on `MachinePool`s  in the hierarchy below.  A scheduler on every 
+layer should assign a `Machine` on a `MachinePool` which can fulfill it.  This proposal addresses the known 
+limitations of the current implementation:
+- API changes needed to extend the scheduling if further resources should be considered
+- Available resources are attached to `Pools` and aggregated. In higher levels a correct decision can't be made 
+  which means that `Machine`s are assigned to saturated `Pool`s.
+
+### Goals
+
+- Dynamic (`NetworkInterfaces`, `LocalDiskStorage`) and static resources (resources defined by a `Class` like `CPU`, 
+  `Memory`) should be taken into the scheduling decision.
+- Scheduling should be robust against data race conditions
+- `Machine`s on every level should be scheduled on a `Pool`
+- It should be possible to add resources influencing the scheduling decision without API change
+- Resource updates of `Pool`s should be possible
+- Adding/removing of compute `Pool`s should be possible
+
+### Non-Goals
+- The scheduler should not act on `Machine` updates like `NetworkInterface` attachments
+- This proposal does not cover the scheduling of `Volumes`.
+
+## Proposal
+
+For every created `Machine` with an empty `spec.machinePoolRef`, the scheduler will create a resource `Reservation`. 
+The `poollet` will continue to broker only `Machine`s with a `.spec.machinePoolRef` set. The `Reservation` is a 
+condensed derivation of a `Machine` containing the requested resources like `NetworkInterfaces` and the attached 
+`MachineClass` resource list. The `IRI` will be extended to be able to manage the `Reservation`s. Once the 
+`Reservation` hits a pool provider (e.g. `libvirt-provider`), the decision can be made if the `Reservation` can be 
+fulfilled or not. In case of accepting the `Reservation`, resources needs to be blocked until: 
+1. the reservation is being deleted 
+2. the corresponding `Machine` is being placed on the pool provider.
+
+The status of the `Reservation` is being propagated up to the layer where it was created. As soon as the root 
```

Comment:
how is it ensured that a scheduler on a lower hierarchy level does not start scheduling / acting upon the update for the particular `reservation` object? the status is consecutively brokered up so an intermediate layer will see an update to the `reservation` before the "root" `reservation` / scheduler.

Reply:
ah forget this, as soon as i hit "add" i realized there is no `machine` object on that intermediate layer because brokering for the `machine` only starts when it has a machinepool reference set.